### PR TITLE
[cc65] Better support for putting stuff on the zeropage

### LINKED ISSUE #243: Indirect array accesses not ZP-optimized
It seems accesses of the type "array[var]" do not get the one-byte (or more) reduction, even when both array and var are in ZP. The direct type, "array[13]", does get it.

All under -Oisr.

### LINKED ISSUE #261: import/export of zeropage symbols using pragmas in C does not work as expected
what i did was exporting a symbol in one C file, using pragmas to make it a zeropage symbol and also pushing it into a custom segment. and then i import it in another file, again using pragma to make it a zeropage symbol (as described in the docs):

``` C
extern char foo;
#pragma zpsym ("foo")
int n;
int main(void) {
    n = foo;
    return 0;
}
```

``` C
//#pragma bss-name (push,"ZEROPAGE")            /* no warning */
#pragma bss-name (push,"MYZEROPAGE")            /* warning */
char foo;
#pragma zpsym ("foo")
#pragma bss-name (pop)
```

now the odd thing is, the assembler gives this warning:

``` C
zpsym2.s(18): Warning: Symbol `foo' is absolute but exported zeropage
```

...which seems wrong (and is a bit irritating). also the generated code is correct (uses zeropage adressing).

whats interesting is that the warning goes away when putting the symbol into the regular "ZEROPAGE" segment (without changing anything else) - that seems to imply that its actually a bug, because for extra zeropage variables using an extra segment is pretty much required.

### LINKED ISSUE #917: [feature request] Option to have global variables in zeropage
Global variables in zeropage can produce more than >10% speed-ups. This can be implemented currently by a cumbersome procedure:

1. declaring then as extern 
2. using the zpsym pragma 
3. and then linking a .s file with their actual location in zero-page

It would be nice to trigger this behavior with a new option (-Oz ?) that would just take the first global variables in the order in which they are declared and map them in zero-page locations that are usually available for C-only projects for each specific target (or use a target-specific config file for this). 

An example of a nice >10% speed-up produced by this optimization is testable in my project:
https://github.com/Fabrizio-Caruso/Primitive-Pythagorean-Triples

Remark: ANSI C does not allow the register qualifier on global variables. So we cannot have this behavior as a result of the register qualifier on global variables.

### PR DESCRIPTION
**Pull request by acqn**
- [X] The instruction parser can now recognize ZP locations and set the addressing mode for them.
- [X] Support for optionally specifying address size for segments with:
`#pragma ***-name([push,] "name"[, "addrsize"])`
where "addrsize" is any addressing size supported in ca65.
- [X] Symbols in ZP segments will use '.exportzp' instead of '.export' if exported.

This resolves #261 and resolves #917 with the new features. It could be considered that this also "almost resolves" #243 (as in the OP) since this opens a path to the expected optimizations, although there are still some other factors that could prevent the optimizations from happening.

The new optional parameter of `#pragma ***-name`
---
`#pragma ***-name([push,] "name"[, "addrsize"])`

Example: https://github.com/cc65/cc65/issues/917#issuecomment-647326244.

The new optional "addrsize" string is to add an address size string for a segment in asm output and in theory can be anything, but only those recognized by ca65 make sense.

- Absolute address size strings
  - "abs"
  - "absolute"
  - "near"
- Default address size strings
  - "default"
- Far address size strings
  - "far"
- Long address size strings
  - "dword"
  - "long"
- Zeropage address size strings
  - "direct"
  - "zeropage"
  - "zp"

Moreover, symbols put on a zeropage specified with this pragma will be exported with `.exportzp` instead of `.export`.

### CODE REVIEW

**Comment by jmr** [comment]
> * [x]  Support for optionally specifying address size for segments with:
>   `#pragma ***-name([push,] "name"[, "addrsize"])`
>   where "addrsize" is any addressing size supported in ca65. 

Is there some `#pragma` documentation in `doc/cc65.sgml` that can be updated?

**Comment by acqn** [comment]
> 
> 
> > * [x]   Support for optionally specifying address size for segments with:
> >   `#pragma ***-name([push,] "name"[, "addrsize"])`
> >   where "addrsize" is any addressing size supported in ca65.
> 
> Is there some `#pragma` documentation in `doc/cc65.sgml` that can be updated?

Yep. About the new optional address size parameter and the change of the behavior of `.export` as well.

**Comment by oliverschmidt** [comment]
>> Is there some #pragma documentation in doc/cc65.sgml that can be updated?

> [...] I can take care of that if no one else voluteers to. (But look out for my English!)

Sorry for the misunderstanding. I sort of implied that your answer would mean that you provide a documentation enhancement. But now I see that it didn't. Therefore I explicitly ask you to provide some docs. I presume @greg-king5 is willing to make sure the new content is correct language-wise.

**Comment by greg-king5** [comment]
> > > Is there some `#pragma` documentation in "doc/cc65.sgml" that can be updated?
> 
> > [...] I can take care of that if no one else volunteers to. (But, look out for my English!)
> 
> Sorry for the misunderstanding. I sort of implied that your answer would mean that you provide a documentation enhancement. But, now I see that it didn't. Therefore, I explicitly ask you to provide some docs. I presume @greg-king5 is willing to make sure the new content is correct language-wise.

Yep. :grin:

File: src/cc65/segments.c [hunk 1]
```diff
@@ -85,6 +96,86 @@ static Collection SegmentStack = STATIC_COLLECTION_INITIALIZER;
...
+
+
+void DoneSegAddrSizes (void)
+/* Free the segment address sizes */
+{
+    SegAddrSize_t* A;
+    int I;
+    for (I = 0; I < (int)CollCount (&SegmentAddrSizes); ++I) {
+        A = CollAtUnchecked (&SegmentAddrSizes, I);
+        SB_Done (A->Name);
```
Comment (greg-king5) [inline src/cc65/segments.c:115]: `SB_Done()` frees the string _inside_ the StrBuf.  That line must free the entire StrBuf; use `FreeStrBuf()`.

**Review by greg-king5 — CHANGES_REQUESTED** [review]

File: src/cc65/codeinfo.c [hunk 2]
```diff
@@ -377,6 +378,85 @@ static const ZPInfo ZPInfoTable[] = {
...
+/* Exam if the main part of the arg string indicates a ZP loc */
+{
+    unsigned short  ArgInfo = 0;
+    long            Offset = 0;
+    StrBuf          NameBuf;
+    SymEntry*       E = 0;
+    const ZPInfo*   Info = 0;
+
+    /* The actual name can never be longer than the arg string */
+    SB_InitFromString (&NameBuf, xstrdup (Name));
```
[the remark this answers is not in the document]
Comment (acqn) [reply src/cc65/codeinfo.c:402]: One more misuse of ` SB_InitFromString` like that in #1277. It doesn't need to pre-allocate the buffer.

File: src/cc65/codeinfo.c [hunk 3]
```diff
@@ -377,6 +378,85 @@ static const ZPInfo ZPInfoTable[] = {
 
 
 
+int IsAbsAddrOnZP (long Address)
+/* Return true if the Address is within the ZP range of the platform.
+** FIXME: ZP range may vary on different platforms.
```
[the remark this answers is not in the document]
Comment (acqn) [reply src/cc65/codeinfo.c:383]: Weird.... The function name should have been ` IsAddrOnZP ` as I recall naming it. The current one seems to be an interim edit. Anyways it's an easy fix.

File: src/cc65/codeinfo.c [hunk 4]
```diff
@@ -377,6 +378,85 @@ static const ZPInfo ZPInfoTable[] = {
...
+            return Offset >= 0 && Offset < (int)Info->Size;
+        }
+
+        /* Assume it be non-ZP */
+        return 0;
+    }
+
+    if ((ArgInfo & AIF_EXTERNAL) == 0) {
+        /* We don't support local variables on ZP */
+        return 0;
```
[the remark this answers is not in the document]
Comment (acqn) [reply src/cc65/codeinfo.c:436]: Theis piece of code was cut & pasted from somewhere outside into this function. Typically missing clean-ups for multiple function exits.

File: src/cc65/segments.c (hunk 1 above)
Comment (acqn) [reply src/cc65/segments.c:115]: Hmmm, `A->Name` was not intended to be allocated separately from `A`.

File: src/cc65/codeinfo.c [hunk 5]
```diff
@@ -40,6 +40,7 @@
 #include "chartype.h"
 #include "coll.h"
 #include "debugflag.h"
+#include "xmalloc.h"
```
Comment (greg-king5) [inline src/cc65/codeinfo.c:43]: That `#include` isn't needed, now.

File: src/cc65/segments.c [hunk 6]
```diff
@@ -85,6 +96,86 @@ static Collection SegmentStack = STATIC_COLLECTION_INITIALIZER;
...
 
+void InitSegAddrSizes (void)
+/* Initialize the segment address sizes */
+{
+    InitCollection (&SegmentAddrSizes);
+}
+
+
+
+
```
Comment (greg-king5) [inline src/cc65/segments.c:107]: One blank line too many.

**Review by greg-king5 — APPROVED** [review]

File: src/cc65/codeinfo.c (hunk 5 above)
Comment (acqn) [reply src/cc65/codeinfo.c]: Removed.

File: src/cc65/segments.c (hunk 6 above)
Comment (acqn) [reply src/cc65/segments.c]: Removed.

File: src/cc65/pragma.c [hunk 7]
```diff
@@ -430,13 +433,35 @@ static void SegNamePragma (StrBuf* B, segment_t Seg)
...
+            SB_Skip (B);
+            SB_SkipWhite (B);
+
+            /* A string argument must follow */
+            if (!GetString (B, &A)) {
+                goto ExitPoint;
+            }
+
+            /* Get the address size string for the segment */
+            AddrSize = AddrSizeFromStr (SB_GetConstBuf (&A));
```
Comment (greg-king5) [inline src/cc65/pragma.c:454]: That expression actually gets the size, not the "size string".  The `GetString()` call, above, got the string.

**Comment by greg-king5** [comment]
The code is good.  There will be no more nitpicking.  This PR will be merged soon after the comment is fixed.

**Comment by greg-king5** [comment]
> The code is good. There will be no more nitpicking. This PR will be merged soon after the comment is fixed.

Oops!  I forgot about "doc/cc65.sgml".  I'll wait for that patch.

**Comment by acqn** [comment]
> 
> 
> > The code is good. There will be no more nitpicking. This PR will be merged soon after the comment is fixed.
> 
> Oops! I forgot about "doc/cc65.sgml". I'll wait for that patch.

Sorry, I am a bit too busy currently to edit that file comforming to the syntax and style by myself. I've added some brief descriptions in the OP about the new features wishing for some volunteers. :p

**Comment by oliverschmidt** [comment]
> @greg-king5: The code is good. There will be no more nitpicking. This PR will be merged soon after the comment is fixed.

> @acqn: Sorry, I am a bit too busy currently to edit that file comforming to the syntax and style by myself.

Based on those two statements it seems best to merge this PR now. I presume that the cc65 community agrees that given the value of contributions by @acqn it wouldn't be appropriate to consider the docs obligatory.

**Comment by mrdudz** [comment]
may i suggest to mark the old "#pragma zpsym" deprecated then and issue a warning that points to the new pragma? that would be very useful, imho.

**Pull request merged**